File: app/src/main/java/com/sysu/edu/api/AccountManager.java
package com.sysu.edu.api;

import android.content.Context;
import android.util.Base64;
import android.util.Pair;

import androidx.datastore.preferences.core.MutablePreferences;
import androidx.datastore.preferences.core.Preferences;
import androidx.datastore.preferences.core.PreferencesKeys;
import androidx.datastore.preferences.rxjava3.RxPreferenceDataStoreBuilder;
import androidx.datastore.rxjava3.RxDataStore;

import com.google.crypto.tink.Aead;
import com.google.crypto.tink.KeyTemplates;
import com.google.crypto.tink.RegistryConfiguration;
import com.google.crypto.tink.aead.AeadConfig;
import com.google.crypto.tink.integration.android.AndroidKeysetManager;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class AccountManager {
    
    private static volatile AccountManager INSTANCE;
    private final RxDataStore<Preferences> dataStore;
    private final Aead aead;
    
    private AccountManager(Context context) {
        try {
            AeadConfig.register();
            aead = new AndroidKeysetManager.Builder()
                    .withSharedPref(context, "master_keyset", "secure_keys")
                    .withKeyTemplate(KeyTemplates.get("AES256_GCM"))
                    .withMasterKeyUri("android-keystore://master_key")
                    .build()
                    .getKeysetHandle()
                    .getPrimitive(RegistryConfiguration.get(), Aead.class);
        } catch (Exception e) {
            throw new RuntimeException("Tink 初始化失败", e);
        }
        dataStore = new RxPreferenceDataStoreBuilder(context, "accounts").build();
    }
    
    public static AccountManager getInstance(Context context) {
        if (INSTANCE == null) {
            synchronized (AccountManager.class) {
                if (INSTANCE == null)
                    INSTANCE = new AccountManager(context.getApplicationContext());
            }
        }
        return INSTANCE;
    }
    
    public Pair<String, String> getActiveAccountSync(String domain) {
        String username = dataStore.data().blockingFirst().get(PreferencesKeys.stringKey("active:" + domain));
        if (username == null) return null;
        String password = getPasswordSync(domain, username);
        return new Pair<>(username, password);
    }
    
    public String getPasswordSync(String domain, String username) {
        String encoded = dataStore.data().blockingFirst().get(PreferencesKeys.stringKey(domain + ":" + username));
        if (encoded == null) return null;
        try {
            return new String(aead.decrypt(Base64.decode(encoded, Base64.DEFAULT), null));
        } catch (Exception e) {
            return null;
        }
    }
    
    public Single<Pair<String, String>> getActiveAccountAsync(String domain) {
        return dataStore.data().firstOrError()
                .map(prefs -> {
                    String username = prefs.get(PreferencesKeys.stringKey("active:" + domain));
                    if (username == null) throw new Exception("No active account");
                    String encoded = prefs.get(PreferencesKeys.stringKey(domain + ":" + username));
                    if (encoded == null) throw new Exception("Password not found");
                    String password = new String(aead.decrypt(Base64.decode(encoded, Base64.DEFAULT), null));
                    System.out.println("Username " + username + " Decrypted password: " + password);
                    return new Pair<>(username, password);
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }
    
    public Completable setAccountAsync(String domain, String username, String password) {
        return Completable.fromAction(() -> dataStore.updateDataAsync(prefs -> {
            MutablePreferences mutable = prefs.toMutablePreferences();
            mutable.set(PreferencesKeys.stringKey(domain + ":" + username), Base64.encodeToString(aead.encrypt(password.getBytes(), null), Base64.DEFAULT));
            return Single.just(mutable);
        })).subscribeOn(Schedulers.io());
    }
    
    public Completable setActiveAccountAsync(String domain, String username) {
        return Completable.fromAction(() -> dataStore.updateDataAsync(prefs -> {
            MutablePreferences mutable = prefs.toMutablePreferences();
            mutable.set(PreferencesKeys.stringKey("active:" + domain), username);
            return Single.just(mutable);
        })).subscribeOn(Schedulers.io());
    }
    
    public Completable removeAccountAsync(String domain, String username) {
        return Completable.fromAction(() -> dataStore.updateDataAsync(prefs -> {
            MutablePreferences mutable = prefs.toMutablePreferences();
            mutable.remove(PreferencesKeys.stringKey(domain + ":" + username));
            return Single.just(mutable);
        })).subscribeOn(Schedulers.io());
    }
    
    public Completable removeActiveAccountAsync(String domain) {
        return Completable.fromAction(() -> dataStore.updateDataAsync(prefs -> {
            MutablePreferences mutable = prefs.toMutablePreferences();
            mutable.remove(PreferencesKeys.stringKey("active:" + domain));
            return Single.just(mutable);
        })).subscribeOn(Schedulers.io());
    }
}

File: app/src/main/java/com/miyuyan/sysuer/api/AccountManager.kt
package com.miyuyan.sysuer.api

import android.content.Context
import android.util.Base64
import android.util.Pair
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.rxjava3.RxPreferenceDataStoreBuilder
import com.google.crypto.tink.Aead
import com.google.crypto.tink.KeyTemplates
import com.google.crypto.tink.RegistryConfiguration
import com.google.crypto.tink.aead.AeadConfig
import com.google.crypto.tink.integration.android.AndroidKeysetManager
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.schedulers.Schedulers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlin.concurrent.Volatile

@OptIn(ExperimentalCoroutinesApi::class)
class AccountManager private constructor(context: Context) {
	private val dataStore = RxPreferenceDataStoreBuilder(context, "accounts").build()
	private val aead: Aead

	init {
		try {
			AeadConfig.register()
			aead = AndroidKeysetManager.Builder()
				.withSharedPref(context, "master_keyset", "secure_keys")
				.withKeyTemplate(KeyTemplates.get("AES256_GCM"))
				.withMasterKeyUri("android-keystore://master_key")
				.build().keysetHandle.getPrimitive(RegistryConfiguration.get(), Aead::class.java)
		} catch (e: Exception) {
			throw RuntimeException("Tink 初始化失败", e)
		}
	}

	/**
	 * 同步获取 domain 域名下的活跃账号
	 * @param domain 域名
	 * @return 活跃账号，<用户名, 密码>
	 * */
	fun getActiveAccountSync(domain: String?): Pair<String?, String?>? {
		val username =
			dataStore.data().blockingFirst()[stringPreferencesKey("active:$domain")] ?: return null
		val password = getPasswordSync(domain, username)
		return Pair<String?, String?>(username, password)
	}

	/**
	 * 同步获取 domain 域名下的账号 username 的密码
	 * @param domain 域名
	 * @param username 用户名
	 * @return 密码
	 * */
	fun getPasswordSync(domain: String?, username: String?): String? {
		val encoded = dataStore.data().blockingFirst()[stringPreferencesKey("$domain:$username")]
			?: return null
		return try {
			String(aead.decrypt(Base64.decode(encoded, Base64.DEFAULT), null))
		} catch (_: Exception) {
			null
		}
	}

	/**
	 * 异步获取 domain 域名下的活跃账号
	 * @param domain 域名
	 * @return 活跃账号，<用户名, 密码>
	 * */
	fun getActiveAccountAsync(domain: String): Single<Pair<String?, String?>> =
		dataStore.data().firstOrError().map<Pair<String?, String?>> { prefs: Preferences ->
			val username = prefs[stringPreferencesKey("active:$domain")]
				?: return@map Pair<String?, String?>("", "")
			val encoded = prefs[stringPreferencesKey("$domain:$username")]
				?: return@map Pair<String?, String?>("", "")
			val password = String(aead.decrypt(Base64.decode(encoded, Base64.DEFAULT), null))
			Pair<String?, String?>(username, password)
		}.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())

	/**
	 * 异步设置账号，不设置为活跃账号
	 * @param domain 域名
	 * @param username 用户名
	 * @param password 密码
	 * */
	fun setAccountAsync(domain: String, username: String, password: String): Completable =
		setAccountAsync(domain, username, password, false)

	/**
	 * 异步设置账号
	 * @param domain 域名
	 * @param username 用户名
	 * @param password 密码
	 * @param active 是否设置为活跃账号
	 * */
	fun setAccountAsync(
		domain: String, username: String, password: String, active: Boolean
	): Completable = dataStore.updateDataAsync { prefs: Preferences ->
		Single.just(prefs.toMutablePreferences().also {
			it[stringPreferencesKey("$domain:$username")] = Base64.encodeToString(
				aead.encrypt(password.toByteArray(), null), Base64.DEFAULT
			)
		}.also {
			if (active || !it.contains(stringPreferencesKey("active:$domain"))) it[stringPreferencesKey(
				"active:$domain"
			)] = username
		})
	}.ignoreElement().subscribeOn(Schedulers.io())

	/**
	 * 异步设置活跃账号
	 * @param domain 域名
	 * @param username 用户名
	 * */
	fun setActiveAccountAsync(domain: String?, username: String?): Completable =
		dataStore.updateDataAsync { prefs: Preferences? ->
			Single.just(
				prefs!!.toMutablePreferences()
					.also { it[stringPreferencesKey("active:$domain")] = username as String })
		}.ignoreElement().subscribeOn(Schedulers.io())

	/**
	 * 异步删除账号
	 * @param domain 域名
	 * @param username 用户名
	 * */
	fun removeAccountAsync(domain: String, username: String): Completable =
		dataStore.updateDataAsync { prefs: Preferences ->
			Single.just(
				prefs.toMutablePreferences()
					.also { it.remove(stringPreferencesKey("$domain:$username")) })
		}.ignoreElement().subscribeOn(Schedulers.io())

	/**
	 * 异步删除活跃账号
	 * @param domain 域名
	 * */
	fun removeActiveAccountAsync(domain: String): Completable =
		dataStore.updateDataAsync { prefs: Preferences ->
			Single.just(
				prefs.toMutablePreferences()
					.also { it.remove(stringPreferencesKey("active:$domain")) })
		}.ignoreElement().subscribeOn(Schedulers.io())


	companion object {
		@Volatile
		private var INSTANCE: AccountManager? = null

		/**
		 * 获取 AccountManager 实例
		 * @param context 上下文对象
		 * */
		fun getInstance(context: Context): AccountManager =
			INSTANCE ?: synchronized(AccountManager::class.java) {
				INSTANCE ?: AccountManager(context.applicationContext).also { INSTANCE = it }
			}
	}
}
